### PR TITLE
WIP: Add all necessary sled trees to zebra-state as defined in RFC0005

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

At least for the near future, zebrad will not have the full test infrastructure
that we need to properly test new code paths that only get called on blocks
created after the sapling network upgrade, such as the `script::Verifier`. As a
result we are likely going to rely heavily on manual testing to verify new
features work as intended. We will then improve the speed with which we can
manually test zebra by backing up a copy of the `zebra-state` sled database
synced up to the sapling activation height, where we can easily reset by wiping
out the current sled database and replacing it with the backed up copy.

This plan however requires makes changes to the sled database format costly.
Each time we add a new tree to the sled database the backed up database must be
deleted and re-created from genesis. To alleviate this issue we need to get
sled as close to the final format as we can as quickly as possible.

## Solution

This change adds the expected sled trees as defined in the RFC. At the time of
writing the precise set of trees needed is a known unknown. Once we've finished
implementing incremental merkle trees however we should be able to determine
the proper format of the rest of the trees needed and include them in this PR,
hopefully bringing us to our final stable sled database format.

As part of this PR I've introduced a set of helper traits for consolidating the
definitions for how our types are stored and retrieved from sled. This should
help improve auditability by isolating related logic to one part of the file,
and ensuring that we use a single definition for the conversion across
`sled_state.rs`.

## Related Issues

- [Tracking Issue](https://github.com/ZcashFoundation/zebra/issues/1049)

## Test Plan

The introduction of `IntoSled` and `FromSled` makes it possible to unit test
our conversion formats in a way that wasn't possible before. Now it should be
relatively trivial to add a set of proptests that test each type that has both
an `IntoSled` and a `FromSled` implementation to verify that the round trip
conversion always returns the correct value.

- [x] Move serialization helpers into a submodule of `sled_state.rs`
- [ ] Add prop tests for each type that implements both `IntoSled` and `FromSled`

## TODOs

- [x]  Add a version number to the path (e.g., ~/.cache/zebra/v1/mainnet/), detect when an older version is present, and delete it

#### Reviews

- [x] rebase on main and force-push, so GitHub's `Commits` and `Files Changed` only show changes from this PR

#### Bugs

- [x] If the block is a genesis block, skip any transaction updates. (Due to a bug in zcashd, genesis block transactions are ignored during validation.)
  - **Note: this fix requires a state version increment, because states containing genesis transactions allow some transactions that `zcashd` rejects.**

#### Incomplete

- [x] For each `TransparentInput::PrevOut { outpoint, .. }` in the transaction's `inputs()`, remove `outpoint` from `utxo_by_output`
- [x] Insert `(transaction_hash, BE32(block_height) || BE32(tx_index))` to `tx_by_hash`
- [x] For each `JoinSplit` description in the transaction, insert `(nullifiers[0],())` and `(nullifiers[1],())` into `sprout_nullifiers`
- [x] For each `Spend` description in the transaction, insert `(nullifier,())` into `sapling_nullifiers`

When we can compute anchors:
- [ ] Add sprout_anchors and sapling_anchors in `FinalizedState`

### CODE REVIEW

**Comment by teor2345** [comment]
> Once we've finished
> implementing incremental merkle trees however we should be able to determine
> the proper format of the rest of the trees needed and include them in this PR,
> hopefully bringing us to our final stable sled database format.

I'm not sure how long our database format will stay the same. It's probably going to be stable for at least 6 months. But future network upgrades could change that. And if we fix a data bug, we might need to rebuild the database from Genesis.

What's the plan for future database format upgrades / bug fixes?

Should we also add some kind of database file or tree path versioning as part of this PR?

Or should we schedule database versioning in a future alpha/beta release, before the first stable?
(Then we'd have more time to write an RFC for database versioning and updates.)

**Comment by hdevalence** [comment]
The plan for future database format changes is to blow away the database and start over.  Because the database is a cache of chain state, we don't lose any information by doing so, and it's much simpler and more reliable than trying to perform a migration.  Adding a version number to the path (e.g., `~/.cache/zebra/v1/mainnet/`) seems like a good idea, since it allows `zebra-state` to detect when an older version is present and delete it.

That said, it would be nice to minimize the frequency of changes to the database format, which I think is the goal of this PR.

**Comment by teor2345** [comment]
>  Adding a version number to the path (e.g., `~/.cache/zebra/v1/mainnet/`) seems like a good idea, since it allows `zebra-state` to detect when an older version is present and delete it.

I have added this task to the TODO list for this PR.

**Comment by teor2345** [comment]
Can Zebra compute sprout and sapling anchors yet?

If not, can we exclude them from the scope of this PR and the first alpha release?
We can open a tracking issue for follow-up.

**Comment by hdevalence** [comment]
I don't think it's important for this PR that we check whether an older version is present and delete it, or implement any version handling at all.  We just need to have a version separator in the path, so that it's possible to implement that in later changes.

I think it would be sufficient to define a
```rust
const SLED_FORMAT_VERSION: u32 = 0;
```
and change `config.rs` so that instead of 
```rust
let path = self.cache_dir.join(net_dir).join("state");
```
we do
```rust
let path = self.cache_dir
    .join("state")
    .join(format!("v{}", SLED_FORMAT_VERSION))
    .join(net_dir);
```
(I changed the order of the path entries in a way that made more sense to me, so that someone who opens the cache directory sees `state/v0/mainnet` or `state/v3/testnet` etc.  I think that this makes more sense than putting `state` last.)

Then, in a later change, we can implement logic that handles different versions, but that work doesn't need to block this change.

**Comment by teor2345** [comment]
> Can Zebra compute sprout and sapling anchors yet?
> 
> If not, can we exclude them from the scope of this PR and the first alpha release?
> We can open a tracking issue for follow-up.

These tracking items are currently in the state tracking issue.

**Comment by teor2345** [comment]
> Then, in a later change, we can implement logic that handles different versions, but that work doesn't need to block this change.

I opened  #1213 for the "delete previous state versions" implementation.

**Comment by teor2345** [comment]
@yaahc can you please rebase on `main` and force-push, so GitHub's `Commits` and `Files Changed` only show changes from this PR?

**Comment by teor2345** [comment]
I've added the following TODOs to this PR, because they impact the state format (or state validity):
- In `commit_finalized_direct`, if the block is a genesis block, skip any transaction updates. (Due to a bug in zcashd, genesis block transactions are ignored during validation.)
  - **Note: this fix requires a state version increment, because states containing genesis transactions allow some transactions that `zcashd` rejects.**

When we can compute anchors:
    - Add `sprout_anchor` and `sapling_anchor` to `QueuedBlock` 
    - Add `sprout_anchors` and `sapling_anchors` to `FinalizedState`

**Comment by teor2345** [comment]
Actually there are a few more changes, see the TODO list in the PR summary for details.

**Comment by hdevalence** [comment]
Picking this up before finishing the request/response implementation because I'm concerned about merge conflicts.

**Comment by hdevalence** [comment]
I rebased this onto `main`, and pushed the original contents as `sled-trees-1`, which can be compared against `sled-trees-2` (pushed onto this branch) here: https://github.com/ZcashFoundation/zebra/compare/sled-trees-1...sled-trees-2

**Comment by hdevalence** [comment]
@teor2345 I noticed that your TODO list included inserting sprout and sapling nullifiers, and indexing each transaction.  I think those items were already included in @yaahc's commits, so I checked them off -- did you think that there was some missing functionality on those three items?

**Comment by teor2345** [comment]
> @teor2345 I noticed that your TODO list included inserting sprout and sapling nullifiers, and indexing each transaction. I think those items were already included in @yaahc's commits, so I checked them off -- did you think that there was some missing functionality on those three items?

I was working off `main`, because there were multiple in-progress PRs.

If they're done in this PR, that's all I wanted to double-check.

**Comment by hdevalence** [comment]
@teor2345 Great, I just wanted to check to make sure that there wasn't an issue identified with the existing code.  I think that all of the TODO items are addressed, except for the ones that are to be done later (when we can compute anchors).

**Comment by teor2345** [comment]
> @teor2345 Great, I just wanted to check to make sure that there wasn't an issue identified with the existing code. I think that all of the TODO items are addressed, except for the ones that are to be done later (when we can compute anchors).

So to be clear, the current database format is v1...
Will we merge this PR, and switch to v2 when we can compute anchors?
Or will we wait until we can compute anchors, finish this PR, and stay on v1?

**Comment by yaahc** [comment]
> Will we merge this PR, and switch to v2 when we can compute anchors?
> Or will we wait until we can compute anchors, finish this PR, and stay on v1?

I think we should do the former

**Comment by teor2345** [comment]
> > Will we merge this PR, and switch to v2 when we can compute anchors?
> > Or will we wait until we can compute anchors, finish this PR, and stay on v1?
> 
> I think we should do the former

@hdevalence what do you think about merging this PR as-is?

File: zebra-client/src/lib.rs
Comment (teor2345) [inline zebra-client/src/lib.rs:11]: This change could go in a separate PR, if you'd like - along with the identical change in zebra-rpc/src/lib.rs

File: zebra-state/src/sled_state.rs
Comment (teor2345) [inline zebra-state/src/sled_state.rs:74]:
```suggestion
            // TODO: sprout and sapling anchors
            debug_stop_at_height: config.debug_stop_at_height.map(block::Height),
```

File: zebra-state/src/sled_state.rs
Comment (teor2345) [inline zebra-state/src/sled_state.rs:102]:
```suggestion
            // TODO: sprout and sapling anchors

```

File: zebra-state/src/sled_state.rs
Comment (teor2345) [inline zebra-state/src/sled_state.rs:250]: ~I think the sprout anchor from the genesis block is also ignored by `zcashd`, but let's confirm that.~

Edit: there are no sprout anchors until block 100, due to the coinbase rules.

**Review by teor2345 — APPROVED** [review]
I have some minor suggestions about comments and PR structure, which shouldn't block merging.

~We also need to work out if the genesis sprout anchor is valid in `zcashd`, but we shouldn't block the PR on finding that out. (I suspect that it isn't, as `zcashd` doesn't do a lot of the normal block processing on genesis blocks.)~

Edit: there are no sprout anchors until block 100, due to the coinbase rules.

File: zebra-state/src/sled_state.rs
Comment (teor2345) [reply zebra-state/src/sled_state.rs:250]: @dconnolly do we only need to store one sprout anchor per block?
Or is every sprout treestate, after every transaction, a valid sprout anchor for future transactions?

Comment (teor2345) [reply zebra-state/src/sled_state.rs:250]: Hypothetically, if there were Sapling anchors at genesis, `zcashd` would ignore them.

I don't know what Zebra should do with both the sprout and sapling edge cases, see PR #1232 for a potential RFC update.

**Pull request merged**
